sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastNestedLoopJoinExecBase.scala

```diff
@@ -774,27 +774,33 @@ abstract class GpuBroadcastNestedLoopJoinExecBase(
       numFirstTableColumns: Int): RDD[ColumnarBatch] = {
     val buildTime = gpuLongMetric(BUILD_TIME)
     val buildDataSize = gpuLongMetric(BUILD_DATA_SIZE)
-    // NOTE: this is a def because we want a brand new `ColumnarBatch` to be returned
-    // per partition (task), since each task is going to be taking ownership
-    // of a columnar batch via `LazySpillableColumnarBatch`.
-    // There are better ways to fix this: https://github.com/NVIDIA/spark-rapids/issues/7642
-    def builtBatch: ColumnarBatch = {
-      makeBuiltBatch(relation, buildTime, buildDataSize)
-    }
     val streamAttributes = streamed.output
     val numOutputRows = gpuLongMetric(NUM_OUTPUT_ROWS)
     val numOutputBatches = gpuLongMetric(NUM_OUTPUT_BATCHES)
     val opTime = gpuLongMetric(OP_TIME)
     val joinTime = gpuLongMetric(JOIN_TIME)
     val nestedLoopJoinType = joinType
     val buildSide = gpuBuildSide
+    val buildSchema = buildPlan.schema
+    val postProjection = buildSidePostProjection
     streamed.executeColumnar().mapPartitions { streamedIter =>
-      val lazyStream = streamedIter.map { cb =>
+      // Will materialize the stream batch before materializing the build batch.
+      val (buildBatch, bufferedStream) = {
+        GpuBroadcastHelper.getBroadcastBuiltBatchAndStreamIter(
+          relation.asInstanceOf[Broadcast[Any]],
+          buildSchema,
+          streamedIter,
+          Some(buildTime),
+          Some(buildDataSize))
+      }
+      val prjBuildBatch = postProjection.map(_(buildBatch)).getOrElse(buildBatch)
```

Comment:
Better to wrap this by closeOnExcept to avoid memory leak when any exception comes up during the project.
```
val prjBuildBatch = closeOnExcept(buildBatch) { _ =>
   postProjection.map(_(buildBatch)).getOrElse(buildBatch)
}
```